pip_run/commands.py

```diff
@@ -15,7 +16,44 @@ def _is_python_arg(item: str):
     to Python and not to pip install.
     """
     path = pathlib.Path(item)
-    return path.is_file() and path.suffix == '.py'
+    # start with the tests which do very minimal IO
+    if not path.exists():
+        return False
+    if path.suffix == ".py":
+        return True
+
+    # now, sniff the first line of the file and return true if it looks like a
+    # python script based on the shebang line
+    return _has_python_shebang(path)
```

Comment:
It's a fair question, and I'm not certain that I chose the correct path. But I think I prefer some flavor of shebang check over most of the other options.

Parsing doesn't work nicely if we ever want different pythons between the `pip-run` runtime and the virtualenv being built. I'm not sure if `pip-run` would ever support that kind of use, but I'd rather not take an approach which conflicts with it -- especially since "requires-python" was the main non-requirements data which was requested in the PEP 722 discussion.

And I'm not aware of a very clear-cut check I could add for requirements and constraints files.

We could do multiple approaches in layers too; I'd be fine with that. e.g.,
```python
# checking the shebang first avoids a slower parse of the whole file
if _has_python_shebang(path):  # defined as it is today
    return True
if _parses_as_python(path):  # ast.parse with an error handler
    return True
return False
```

If we can define `_parses_as_requirements_txt` and `_parses_as_constraints_txt`, we can use them to work out a policy and ordering of checks. But I'm not sure what we'd do for those other than `shlex` each line, strip comments, try to process and remove some known pip arguments like `--index-url`, and then run a PEP 508 parser on the results. All of that would be fine... except for handling arbitrary `pip install` options -- that part feels dicey to me.

Reply:
For now, we're targeting shebang detection only. Let's see how that goes.